File: Senior/2020/S3.py
def createCounts():
    letters = 'abcdefghijklmnopqrstuvwxyz'
    return {letter: 0 for letter in letters}

needle = input()
haystack = input()

needleLen = len(needle)
haystackLen = len(haystack)

# if the length of needle is longer than haystack, print 0
if needleLen > haystackLen:
    print(0)
else:
    # find counts of each letter in the needle
    needle_counts = createCounts()
    for letter in needle:
        needle_counts[letter] += 1

    results = set()
    counts = createCounts()

    for i in range(needleLen):
        counts[haystack[i]] += 1

    if counts == needle_counts:
        results.add(hash(haystack[0:needleLen]))

    counts[haystack[0]] -= 1

    count = 0

    # sliding window
    for i in range(1, haystackLen - needleLen + 1):
        counts[haystack[i+needleLen-1]] += 1
        if counts == needle_counts:
            results.add(hash(haystack[i:i+needleLen]))
        counts[haystack[i]] -= 1

    print(len(results))

File: Waterloo CCC/Senior/2020/S3.cpp
#include <iostream>
#include <string>
#include <map>
#include <set>

using namespace std;

bool isPerm(map<char, int>& m1, map<char, int>& m2) {
    for (int i = 0; i < 26; i++) {
        char letter = 'a' + i;
        if (m1[letter] != m2[letter]) {
            return false;
        }
    }
    return true;
}

int main() {
    string needle, haystack;
    cin >> needle >> haystack;
    int needle_size = needle.size();
    int haystack_size = haystack.size();

    int count = 0;

    if (needle_size > haystack_size) { // if needle is bigger than haystack, no need to search
        cout << 0 << endl;
        return 0;
    }

    map<char, int> needle_map;
    map<char, int> haystack_map;

    // add all letters of the alphabet to the map
    for (int i = 0; i < 26; i++) {
        needle_map[(char)('a' + i)] = 0;
        haystack_map[(char)('a' + i)] = 0;
    }

    // calculate the frequency of each letter in the needle
    for (int i = 0; i < needle_size; i++) {
        needle_map[needle[i]]++;
    }

    // calculate the frequency of the first needle_size letters of the haystack
    for (int i = 0; i < needle_size; i++) {
        haystack_map[haystack[i]]++;
    }

    set<size_t> found_positions;
    hash<string> str_hash;

    // check if the frequency of the first needle_size letters of the haystack is equal to the frequency of the needle
    if (needle_map == haystack_map) {
        found_positions.insert(str_hash(haystack.substr(0, needle_size)));
        count++;
    }

    // sliding window to check if the frequency of the next letters of the haystack is equal to the frequency of the needle
    for (int i = needle_size; i < haystack_size; i++) {
        haystack_map[haystack[i]]++;
        haystack_map[haystack[i - needle_size]]--;
        if (isPerm(needle_map, haystack_map) && found_positions.find(str_hash(haystack.substr(i - needle_size + 1, needle_size))) == found_positions.end()) {
            found_positions.insert(str_hash(haystack.substr(i - needle_size + 1, needle_size)));
            count++;
        }
    }

    cout << count << endl;
}